python/ray/llm/_internal/batch/stages/base.py

```diff
@@ -0,0 +1,172 @@
+"""TBA"""
+import logging
+from typing import Any, Dict, AsyncIterator, List, Callable
+
+import pyarrow
+from pydantic import BaseModel, Field
+from ray.data.block import UserDefinedFunction
+
+logger = logging.getLogger(__name__)
+
+
+def wrap_preprocess(
+    fn: UserDefinedFunction,
+    processor_data_column: str,
+) -> Callable:
+    """Wrap the preprocess function, so that the output schema of the
+    preprocess is normalized to {processor_data_column: fn(row), other input columns}.
+
+    Args:
+        fn: The function to be applied.
+        processor_data_column: The internal data column name of the processor.
+
+    Returns:
+        The wrapped function.
+    """
+
+    def _preprocess(row: dict[str, Any]) -> dict[str, Any]:
+        # First put everything into processor_data_column.
+        outputs = {processor_data_column: row}
+
+        # Then apply the preprocess function and add its outputs.
+        preprocess_output = fn(row)
+        outputs[processor_data_column].update(preprocess_output)
+        return outputs
+
+    return _preprocess
+
+
+def wrap_postprocess(
+    fn: UserDefinedFunction,
+    processor_data_column: str,
+) -> Callable:
+    """Wrap the postprocess function to remove the processor_data_column.
+    Note that we fuly rely on users to determine which columns to carry over.
+
+    Args:
+        fn: The function to be applied.
+        processor_data_column: The internal data column name of the processor.
+
+    Returns:
+        The wrapped function.
+    """
+
+    def _postprocess(row: dict[str, Any]) -> dict[str, Any]:
+        if processor_data_column not in row:
+            raise ValueError(
+                f"[Internal] {processor_data_column} not found in row {row}"
+            )
+
+        return fn(row[processor_data_column])
+
+    return _postprocess
+
+
+class StatefulStageUDF:
+    def __init__(self, data_column: str):
+        self.data_column = data_column
```

Comment:
I don't understand why the UDF should be have dynamic data_column as its internal state?

Reply:
I mean why not make it a class attribute